### PR TITLE
RD-2965 + RD-3117 Refactor starting executions

### PR DESCRIPTION
**Pull request by tehasdf**
Make it cleaner what parts of starting an execution are under a
transaction, and what must not be.

This should do away with the deadlocks when starting/dequeueing
(note: I haven't touched update-dep-status just yet).

This is done by splitting rendering an execution, from actually sending
it. Rendering can be done on any execution, in a transaction, even if the
execution is not committed yet. Sending is sending.

Commit messages go into more depth on that.

### CODE REVIEW

File: rest-service/manager_rest/resource_manager.py [hunk 1]
```diff
@@ -219,34 +219,29 @@ def start_queued_executions(self, deployment_storage_id):
...
+                                _, messages = self._execute_system_workflow(
+                                    execution, queue=True)
+                            else:
+                                messages = self.prepare_executions(
+                                    [execution], queue=True)
+                            to_run.extend(messages)
+                        except Exception as e:
+                            current_app.logger.warning(
+                                'Could not dequeue execution %s: %s',
+                                execution, e)
```
Comment (tehasdf) [inline rest-service/manager_rest/resource_manager.py:239]: this is ugly, but RD-3163 is about refactoring this

File: rest-service/manager_rest/resource_manager.py [hunk 2]
```diff
@@ -1282,49 +1276,56 @@ def cancel_execution(self, execution_id, force=False, kill=False):
...
+                    (not force_execution
+                     or execution.status != ExecutionState.CANCELLING)\
+                    and not kill_execution:
+                raise manager_exceptions.IllegalActionError(
+                    "Can't {0} cancel execution {1} because it's in status {2}"
+                    .format(
+                        'kill-' if kill_execution
+                        else 'force-' if force_execution else '',
+                        execution_id,
+                        execution.status))
```
Comment (tehasdf) [inline rest-service/manager_rest/resource_manager.py:1308]: RD-3161 is about refactoring this

File: rest-service/manager_rest/resource_manager.py [hunk 3]
```diff
@@ -1059,25 +1048,34 @@ def check_for_executions(self, execution, force, queue):
...
+            return True
+        if force:
             return system_exec_running
-        else:
-            execution_running = self._check_for_active_executions(
-                execution, queue)
-            return system_exec_running or execution_running
+        if not execution.deployment or not execution.deployment._storage_id:
+            return system_exec_running
+        return self._check_for_active_executions(execution, queue)
```
Comment (tehasdf) [inline rest-service/manager_rest/resource_manager.py:1057]: I still hate all the running-checks. RD-3164 is about simplifying them.

File: rest-service/manager_rest/resource_manager.py [hunk 4]
```diff
@@ -1178,15 +1176,11 @@ def _execute_system_workflow(self,
...
-            bypass_maintenance=bypass_maintenance,
-            handler=send_handler,
-        )
-
-        return execution
+            return execution, []
+        message = execution.render_message(
+            bypass_maintenance=bypass_maintenance)
+        db.session.commit()
+        return execution, [message]
```
Comment (tehasdf) [inline rest-service/manager_rest/resource_manager.py:1183]: `_execute_system_workflow` should be not a thing of its own, but it should just use `prepare_executions` as other executions do. RD-3162 is about that.

File: rest-service/manager_rest/upload_manager.py [hunk 5]
```diff
@@ -582,13 +583,14 @@ def _prepare_and_process_doc(self, data_id, blueprint_url,
...
 
         try:
-            temp_blueprint.upload_execution = rm.upload_blueprint(
+            temp_blueprint.upload_execution, messages = rm.upload_blueprint(
                 data_id,
                 application_file_name,
                 blueprint_url,
                 config.instance.file_server_root,   # for the import resolver
                 validate_only=True,
             )
```
Comment (geokala) [inline rest-service/manager_rest/upload_manager.py:592]: Unrelated to this set of commits (I think), but why does this one not have any rm.sm.update or similar?

Comment (tehasdf) [reply rest-service/manager_rest/upload_manager.py:592]: Just ORM Things™

File: rest-service/manager_rest/resource_manager.py [hunk 6]
```diff
@@ -1081,17 +1081,24 @@ def check_for_executions(self, execution, force, queue):
             return system_exec_running or execution_running
 
     def _check_for_active_executions(self, execution, queue):
+        if execution.created_at is not None:
+            status_filter = sql_or(
+                modelsExecution.status.in_(ExecutionState.ACTIVE_STATES),
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py]: modelsExecution - should this be models.Execution?

Comment (tehasdf) [reply rest-service/manager_rest/resource_manager.py]: yeah some later commit fixes this, sorry. I thought i caught all of these in the rebase, but apparently not.

File: rest-service/manager_rest/resource_manager.py [hunk 7]
```diff
@@ -1022,6 +1025,8 @@ def prepare_executions(self, executions, *, force=False, queue=False,
             executions.extend(component_executions)
         if commit:
             db.session.commit()
+        if errors:
+            raise errors[0]
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:1031]: Do we already have a jira for making it raise a list (at least in user, if not machine, readable form)?

Comment (tehasdf) [reply rest-service/manager_rest/resource_manager.py:1022]: This is RD-3166

**Review by geokala — APPROVED** [review]

**Pull request merged**